pkg/scheduler/scheduler.go

```diff
@@ -23,27 +23,34 @@ import (
 	"k8s.io/utils/clock"
 )
 
-// For mocking purposes.
-// This little bit of wrapping needs to be done because go doesn't do
-// covariance, but it does coerce *time.Timer into stoppable implicitly if we
-// write it out like so.
-func afterFunc(c clock.Clock, d time.Duration, f func()) stoppable {
+// We are writting our own time.AfterFunc to be able to mock the clock. The
+// cancel function can be called concurrently.
+func afterFunc(c clock.Clock, d time.Duration, f func()) (cancel func()) {
 	t := c.NewTimer(d)
+	cancelCh := make(chan struct{})
+	cancelOnce := sync.Once{}
+	cancel = func() {
+		t.Stop()
+		cancelOnce.Do(func() {
+			close(cancelCh)
+		})
+	}
 
 	go func() {
-		defer t.Stop()
-		if ti := <-t.C(); ti == (time.Time{}) {
+		defer cancel()
+
+		select {
+		case <-t.C():
+			// We don't need to check whether the channel has returned a zero
```

Comment:
Nit: arguably this comment only makes sense if one knows what code was there before.

Reply:
Good point.

Since I still think that `t.Stop` not closing `t.C` is an important gotcha, I'll move the comment to above the `cancelCh` declaration to remind us of "why" we can't expect `t.Stop` to close `t.C`, and why we need this extra `cancelCh` channel:

```go
    // One might think that t.Stop would close the t.C channel, letting us stop
	// the below goroutine with t.Stop. But t.C is actually never closed. We
	// thus need to create our own "cancel" channel.
    cancelCh := make(chan struct{})
	cancelOnce := sync.Once{}
	cancel = func() {
		t.Stop()
		cancelOnce.Do(func() {
			close(cancelCh)
		})
	}
```

What do you think?

Reply:
Agree that it is good to add a comment as to why the second case is needed. Perhaps it would be good to make it so that the reader in this function can understand what happens at higher level:
```
// We need to ensure that when the caller of this function calls `cancel` the goroutine is exited. `t.Stop` won't close the channel so we need the cancel channel.
```

Although you comment is good too.

Reply:
I mixed both comments:

```go
// The caller expects `cancel` to stop the goroutine. Since `t.C` is not
// closed after calling `t.Stop`, we need to create our own `cancelCh`
// channel to stop the goroutine.
```